tests/basictests.cpp

```diff
@@ -53,514 +94,787 @@ bool number_test_small_integers() {
-    printf("Could not parse '%s': %s\n", json.data(), simdjson::error_message(error));
-    return false;
-  }
-  simdjson::document::iterator iter(doc);
-  if(!iter.is_object()) {
-    printf("Root should be object\n");
-    return false;
-  }
-  if(iter.move_to_key("bad key")) {
-    printf("We should not move to a non-existing key\n");
-    return false;    
-  }
-  if(!iter.is_object()) {
-    printf("We should have remained at the object.\n");
-    return false;
-  }
-  if(iter.move_to_key_insensitive("bad key")) {
-    printf("We should not move to a non-existing key\n");
-    return false;    
-  }
-  if(!iter.is_object()) {
-    printf("We should have remained at the object.\n");
-    return false;
-  }
-  if(iter.move_to_key("bad key", 7)) {
-    printf("We should not move to a non-existing key\n");
-    return false;    
-  }
-  if(!iter.is_object()) {
-    printf("We should have remained at the object.\n");
-    return false;
-  }
-  if(!iter.down()) {
-    printf("Root should not be emtpy\n");
-    return false;
-  }
-  if(!iter.is_string()) {
-    printf("Object should start with string key\n");
-    return false;
+namespace parse_api_tests {
+  using namespace std;
+  using namespace simdjson;
+
+  const padded_string BASIC_JSON = string("[1,2,3]");
+  const padded_string BASIC_NDJSON = string("[1,2,3]\n[4,5,6]");
+  // const padded_string EMPTY_NDJSON = string("");
+
+  bool document_parse() {
+    std::cout << "Running " << __func__ << std::endl;
+    auto [doc, error] = document::parse(BASIC_JSON);
+    if (error) { cerr << error << endl; return false; }
+    if (!doc.root().is_array()) { cerr << "Document did not parse as an array" << endl; return false; }
+    return true;
   }
-  if(iter.prev()) {
-    printf("We should not be able to go back from the start of the scope.\n");
-    return false;
+  bool parser_parse() {
+    std::cout << "Running " << __func__ << std::endl;
+    document::parser parser;
+    auto [doc, error] = parser.parse(BASIC_JSON);
+    if (error) { cerr << error << endl; return false; }
+    if (!doc.root().is_array()) { cerr << "Document did not parse as an array" << endl; return false; }
+    return true;
   }
-  if(strcmp(iter.get_string(),"Image")!=0) {
-    printf("There should be a single key, image.\n");
-    return false;
+  bool parser_parse_many() {
+    std::cout << "Running " << __func__ << std::endl;
+    document::parser parser;
+    int count = 0;
+    for (auto [doc, error] : parser.parse_many(BASIC_NDJSON)) {
+      if (error) { cerr << error << endl; return false; }
+      if (!doc.root().is_array()) { cerr << "Document did not parse as an array" << endl; return false; }
+      count++;
+    }
+    if (count != 2) { cerr << "parse_many returned " << count << " documents, expected 2" << endl; return false; }
+    return true;
   }
-  iter.move_to_value();
-  if(!iter.is_object()) {
-    printf("Value of image should be object\n");
-    return false;
+  // bool parser_parse_many_empty() {
```

Comment:
Is parser_parse_many_empty mean to be commented out?

Reply:
Yep. I wanted to add the test while I was in there if it was already succeeding, but it fails. It's not a new failure, though. Either there is a bug, or my expectations are wrong (I expect it to succeed, but not yield any documents).